File: 2025-04-07/forkjoin/src/test/java/LongestSequenceTest.java
import org.junit.Test;

import static org.junit.Assert.assertEquals;
import java.util.Random;


public class LongestSequenceTest {

    private int sequentialCount(int item, int[] arr) {
        int greatestCount = 0;
        int localCount = 0;
        for (int i=0; i < arr.length; i++) {
            if (LongestSequenceForkJoin.compare(arr[i], item)) {
                localCount++;
                if (localCount > greatestCount) {
                    greatestCount = localCount;
                }
            } else {
                localCount = 0;
            }
        }
        return greatestCount;
    }

    @Test
    public void testTrivialLongestSequence() throws Exception {
        int[] arr = {2, 17, 17, 8, 17, 17, 17, 0, 17, 1};
        int runSize = LongestSequenceForkJoin.longestSequence(17, arr, 100);
        assertEquals(3, runSize);
    }

    @Test
    public void testLongestSequence() throws Exception {
        Random r = new Random(90125);
        int[] arr = new int[10_000_000];
        for (int i = 0; i < arr.length; i++) {
            arr[i] = r.nextInt(3);
        }
        int lookupValue = 2;

        // Do the sequential test a number of times to warm it up, measure on
        // last
        int numWarmups = 3;
        long timeBefore = 0;
        long seqTime = 0;
        int seqCount = 0;
        for (int i=0; i < numWarmups; i++) {
            timeBefore = System.currentTimeMillis();
            seqCount = sequentialCount(lookupValue, arr);
            seqTime = System.currentTimeMillis() - timeBefore;
        }

        int parallelCount = 0;
        long parallelTime = 0;

        // Do the parallel test a number of times to warm it up, measure on last
        // int[] seqCutoffs = {1, 2, 10, 100, 1000, 10000};
        int[] seqCutoffs = {
            1, 2, 3, 4, 10, 100, 1000, 10000, 100_000, 1_000_000, 5_000_000
        };
        for (int seqCutoff : seqCutoffs) {
            for (int i=0; i < numWarmups; i++) {
                timeBefore = System.currentTimeMillis();
                parallelCount = LongestSequenceForkJoin
                    .longestSequence(lookupValue, arr, seqCutoff);
                parallelTime = System.currentTimeMillis() - timeBefore;
            }
            assertEquals(seqCount, parallelCount, 1e-9);
            double speedup = ((double) seqTime) / parallelTime;
            System.out.printf("%10d %5d %5d %7.3f\n", seqCutoff, seqTime,
                              parallelTime, speedup);
        }
    }
}

File: 2025-04-07/forkjoin/src/test/kotlin/LongestSequenceTest.kt
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertTrue
import kotlin.random.Random
import kotlin.time.measureTimedValue


class LongestSequenceTest {

    fun sequentialCount(arr: List<Int>): Int {
        var greatestCount = 0
        var localCount = 0
        for (i in 0..<arr.count()) {
            if (isPrime(arr[i])) {
                localCount++
                if (localCount > greatestCount) {
                    greatestCount = localCount
                }
            } else {
                localCount = 0
            }
        }
        return greatestCount
    }

    @Test
    fun testTrivialLongestSequence() {
        val arr = listOf(2, 18, 11, 7, 17, 12, 10, 23, 14, 11)
        val runSize = longestSequence(arr, 100)
        assertEquals(3, runSize)
    }

    @Test
    fun testLongestSequence() {
        var r = Random(90125)
        val arr = mutableListOf<Int>()
        for (i in 0..<50_000) {
            // Make all of the numbers odd; gives them a better chance of being
            // prime
            arr.add((r.nextInt(400_000)*2)+1)
        }

        // Do the sequential test a number of times to warm it up, then do it
        // one more time and measure on last
        val numWarmups = 3
        var total = 0
        for (i in 0..<numWarmups) {
            total += sequentialCount(arr)
        }
        val (seqResult, seqTime) = measureTimedValue {
            sequentialCount(arr)
        }
        // Print the warmup total just to make sure that the code
        // doesn't get optimized out
        println("Warmup sequential total = ${total}")
        println("Max sequence length = ${seqResult}")
        println("Measured sequential time = ${seqTime}")

        val numProcessors = Runtime.getRuntime().availableProcessors()
        println("Number of processors = ${numProcessors}")
        println("-----------------------------")
        val seqCutoffs = listOf(
            1, 2, 3, 4, 10
        )
        println("cutoff / seq time / parallel time / speedup")

        for (seqCutoff in seqCutoffs) {
            // Do the parallel test a number of times to warm it up, measure on
            // last
            total = 0
            for (i in 0..<numWarmups) {
                total += longestSequence(arr, seqCutoff)
            }
            val (parallelResult, parallelTime) = measureTimedValue {
                longestSequence(arr, seqCutoff)
            }

            val speedup = seqTime/parallelTime
            println(String.format("%10d %s %s %7.3f", seqCutoff, seqTime,
                                  parallelTime, speedup))
            assertEquals(seqResult, parallelResult)
        }
    }
}
